File: tools/sec1_audit.py
#!/usr/bin/env python3
"""E.SEC1 read-only historical audit.

Reports COUNTS, KINDS and record identifiers ONLY. It never prints a matching value, not even
truncated, and it never writes to anything it reads. The decision logs are hash-chained, so this
opens them read-only and proposes nothing.

The detection mirrors crates/mind-types/src/safety.rs. It is a second implementation on purpose:
if the two disagree the audit is wrong in a visible way rather than silently agreeing with a bug.
"""
import json, os, re, sqlite3, sys, collections

TOKEN_SHAPES = [("ghp_", 8, False), ("gho_", 8, False), ("ghu_", 8, False), ("ghs_", 8, False),
                ("github_pat_", 8, False), ("glpat-", 8, False), ("xoxb-", 8, False),
                ("xoxp-", 8, False), ("sk-", 6, False), ("akia", 16, True), ("asia", 16, True)]
CARD_CTX = {"card", "cards", "pin", "pins", "cvv", "cvc", "iban", "pan"}
SSN_CTX = {"ssn", "social"}
PHRASES = ["password", "passcode", "passphrase", "api key", "apikey", "api-key", "secret key",
           "access token", "auth token", "refresh token", "client secret", "bearer", "private key"]
TOKEN_RE = re.compile(r"[A-Za-z0-9_-]+")

def luhn(d):
    s = 0
    for i, ch in enumerate(reversed(d)):
        n = int(ch)
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9
        s += n
    return s % 10 == 0

def digit_runs(text, want_free=False):
    """Digit runs. With want_free, only runs NOT buried inside a longer alphanumeric token — a
    64-char hex hash contains card-shaped substrings by chance (28 of 11,866 lines on the box)."""
    out, i, n = [], 0, len(text)
    while i < n:
        if not text[i].isdigit():
            i += 1
            continue
        start = i
        digits = ""
        while i < n and (text[i].isdigit() or text[i] in " -"):
            if text[i].isdigit():
                digits += text[i]
            i += 1
        embedded = (start > 0 and text[start - 1].isalpha()) or (i < n and text[i].isalpha())
        if not (want_free and embedded):
            out.append(digits)
    return out

def value_follows(after):
    for tok in TOKEN_RE.findall(after[:48])[:3]:
        if len(tok) >= 6 and (any(c.isdigit() for c in tok) or len(tok) >= 12):
            return True
    return False

def kinds(text):
    """Every KIND present. Never returns any part of the text."""
    if not text:
        return set()
    low, found = text.lower(), set()
    if "-----begin" in low and "private key" in low:
        found.add("pem-private-key")
    toks = TOKEN_RE.findall(text)
    for tok in toks:
        for pre, min_tail, upper in TOKEN_SHAPES:
            if tok.lower().startswith(pre) and len(tok) - len(pre) >= min_tail:
                if not upper or all(c.isupper() or c.isdigit() for c in tok):
                    found.add("token")
    runs = digit_runs(text)
    for d in digit_runs(text, want_free=True):
        if 13 <= len(d) <= 19 and d[0] in "3456" and luhn(d):
            found.add("payment-card")
    low_toks = {t.lower() for t in toks}
    if low_toks & CARD_CTX and any(len(d) >= 4 for d in runs):
        found.add("card-context-number")
    if low_toks & SSN_CTX and any(len(d) == 9 for d in runs):
        found.add("national-id")
    for p in PHRASES:
        at = low.find(p)
        while at != -1:
            before_ok = at == 0 or not (low[at - 1].isalnum() or low[at - 1] in "_-")
            if before_ok and value_follows(text[at + len(p):]):
                found.add("credential-phrase")
                break
            at = low.find(p, at + len(p))
    return found

def kinds_of_values(node):
    """Every KIND across a parsed JSON document's STRING values, each scanned on its own.

    Numbers are rendered back to text individually, so two adjacent fields can never be read as one
    long digit run — the artifact that a raw-line scan produces and then reports as a payment card.
    """
    found = set()
    if isinstance(node, str):
        return kinds(node)
    if isinstance(node, bool) or node is None:
        return found
    if isinstance(node, (int, float)):
        return kinds(str(node))
    if isinstance(node, list):
        for item in node:
            found |= kinds_of_values(item)
        return found
    if isinstance(node, dict):
        for k, v in node.items():
            found |= kinds_of_values(v)
        return found
    return found

def audit_jsonl(path):
    per_kind, hits, total = collections.Counter(), [], 0
    with open(path, "r", encoding="utf-8", errors="replace") as f:
        for lineno, line in enumerate(f, 1):
            if not line.strip():
                continue
            total += 1
            try:
                ev = json.loads(line).get("event", {})
            except Exception:
                continue
            found = set()
            for field in ("goal", "outcome", "predicted", "lesson", "trigger", "chosen", "subject"):
                found |= kinds(ev.get(field) or "")
            for arr in ("candidates", "rejected", "policy", "evidence_ids"):
                for item in ev.get(arr) or []:
                    found |= kinds(item or "")
            if found:
                for k in found:
                    per_kind[k] += 1
                hits.append((lineno, ev.get("event_id") or f"line:{lineno}", sorted(found)))
    return total, per_kind, hits

def audit_db(path):
    per_kind, hits, total = collections.Counter(), [], 0
    try:
        con = sqlite3.connect(f"file:{path}?mode=ro", uri=True)
        cur = con.execute("SELECT rid, text FROM memories")
    except Exception as e:
        return None, f"{type(e).__name__}", []
    for rid, text in cur:
        total += 1
        found = kinds(text or "")
        if found:
            for k in found:
                per_kind[k] += 1
            hits.append((rid, sorted(found)))
    con.close()
    return total, per_kind, hits

print("E.SEC1 READ-ONLY HISTORICAL AUDIT — counts, kinds and record ids only; no values are printed")
print()
for path in sys.argv[1:]:
    if not os.path.exists(path):
        print(f"{path}: absent")
        continue
    if path.endswith(".jsonl"):
        # A DecisionEvent-shaped log is scanned field by field. Any OTHER jsonl is scanned as RAW
        # LINES, because assuming a schema that does not apply would report a clean bill of health
        # for a file this audit never actually looked inside.
        with open(path, "r", encoding="utf-8", errors="replace") as probe:
            first = probe.readline()
        decision_shaped = '"event"' in first and '"chain"' in first
        if not decision_shaped:
            per_kind, hits, total = collections.Counter(), [], 0
            with open(path, "r", encoding="utf-8", errors="replace") as f:
                for lineno, line in enumerate(f, 1):
                    if not line.strip():
                        continue
                    total += 1
                    # FIELD-AWARE, not raw. Scanning the whole line lets a digit run continue
                    # across JSON punctuation and MANUFACTURE a card-shaped number out of two
                    # unrelated fields — which is what a first raw-line pass appeared to find. Walk
                    # the parsed values instead, and fall back to the raw line only when the line
                    # will not parse at all (which is itself worth seeing).
                    found = set()
                    try:
                        found = kinds_of_values(json.loads(line))
                    except Exception:
                        found = kinds(line)
                    if found:
                        for k in found:
                            per_kind[k] += 1
                        hits.append((lineno, f"line:{lineno}", sorted(found)))
            print(f"{path}  [FIELD-AWARE SCAN of a non-DecisionEvent log]")
            print(f"  lines scanned  : {total}")
            print(f"  lines flagged  : {len(hits)}")
            for k, n in sorted(per_kind.items()):
                print(f"    {k:22} {n}")
            for lineno, eid, ks in hits[:20]:
                print(f"    line {lineno:<6} kinds={','.join(ks)}")
            if len(hits) > 20:
                print(f"    … {len(hits)-20} more flagged lines not listed")
            print()
            continue
        total, per_kind, hits = audit_jsonl(path)
        print(f"{path}\n  events scanned : {total}")
        print(f"  events flagged : {len(hits)}")
        for k, n in sorted(per_kind.items()):
            print(f"    {k:22} {n}")
        for lineno, eid, ks in hits[:20]:
            print(f"    line {lineno:<6} id={eid:<44} kinds={','.join(ks)}")
        if len(hits) > 20:
            print(f"    … {len(hits)-20} more flagged records not listed")
    else:
        total, per_kind, hits = audit_db(path)
        if total is None:
            print(f"{path}\n  unreadable: {per_kind}")
            continue
        print(f"{path}\n  memories scanned : {total}")
        print(f"  memories flagged : {len(hits)}")
        for k, n in sorted(per_kind.items()):
            print(f"    {k:22} {n}")
        for rid, ks in hits[:20]:
            print(f"    rid={rid} kinds={','.join(ks)}")
        if len(hits) > 20:
            print(f"    … {len(hits)-20} more flagged records not listed")
    print()

File: crates/mind-evals/src/bin/sec1_audit.rs
//! E.SEC1b — the read-only historical audit, with production's own detector.
//!
//! Replaces `tools/sec1_audit.py`, which reimplemented the classification rules in a second
//! language. Codex's objection was exact: a duplicated detector can certify a policy production
//! does not run. Everything here classifies through `mind_types` — the same functions guarding the
//! memory write gate, the flight recorder's redaction, the egress tripwire and the eval export.
//! The walking of files and rows is orchestration; none of it decides what is sensitive.
//!
//! REPORTS COUNTS, KINDS AND RECORD IDENTIFIERS ONLY. No matching value is printed, not truncated,
//! not as an offset into content nobody may read. It opens everything read-only and writes nothing:
//! the decision logs are hash-chained and remediation is a separate, preregistered exercise.
//!
//! THE SENSITIVITY CONTROL. Deleting a second implementation costs something real — the Python twin
//! found the 28 chain-hash false positives — so before scanning anything this runs canaries that
//! MUST be flagged and clean strings that must NOT be. If the control fails it exits **2 =
//! INCONCLUSIVE** rather than reporting a clean bill of health. A negative result must demonstrate
//! its own sensitivity, in the same run, on the same build: an instrument that cannot fail is the
//! same defect as a detector that cannot fire.
//!
//! Exit codes: 0 clean · 1 findings (with `--fail-closed`) · 2 the instrument could not prove itself.

use std::collections::BTreeMap;

/// Known-sensitive inputs, one per kind the detector names. Every one must be flagged.
const CANARIES: &[&str] = &[
    "my password is hunter2",
    "ghp_SECRET12345",
    "-----BEGIN RSA PRIVATE KEY-----",
    "my card is 4471 9302 1122 8890",
    "ssn 123456789",
    "AKIAIOSFODNN7EXAMPLE",
];

/// Known-clean inputs. Every one must pass. These are the false positives the old detector had —
/// it refused "asian food recipes" while admitting a password — so they are the control's control.
const CLEAN: &[&str] = &[
    "asian food recipes",
    "the task list for Tuesday",
    "how do passwords work?",
    "where is order 100000000000",
    "what happened at 1756170000000",
];

fn control_passes() -> Result<(), String> {
    for c in CANARIES {
        if mind_types::first_sensitive(c).is_none() {
            return Err(format!("a canary was NOT flagged ({} chars) — the detector cannot fire", c.len()));
        }
    }
    for c in CLEAN {
        if let Some(f) = mind_types::first_sensitive(c) {
            return Err(format!("a clean control WAS flagged as {} — the detector cannot pass", f.kind.label()));
        }
    }
    // The pair rule, which the value-only walk depends on.
    if mind_types::sensitive_pair("api_key", "9f2b1c4d8e").is_none() {
        return Err("the key/value rule did not fire on a named key with a token value".into());
    }
    if mind_types::sensitive_pair("note", "9f2b1c4d8e").is_some() {
        return Err("the key/value rule fired on an unnamed key".into());
    }
    Ok(())
}

/// WHERE a finding is and WHAT it is — never what it says.
///
/// `path` is the JSON key path and `len` the byte length of the matched span. Both are metadata
/// about a value, not the value: this is how the 28 chain-hash false positives were identified
/// (key path plus digit-run length) without anyone reading one.
#[derive(Clone)]
struct Hit {
    path: String,
    kind: &'static str,
    len: usize,
}

/// Every kind present in one scalar. Never returns any part of it.
fn kinds_of(text: &str) -> Vec<&'static str> {
    let mut k: Vec<&'static str> = mind_types::sensitive_findings(text).iter().map(|f| f.kind.label()).collect();
    k.sort_unstable();
    k.dedup();
    k
}

fn hits_of(path: &str, text: &str) -> Vec<Hit> {
    mind_types::sensitive_findings(text)
        .iter()
        .map(|f| Hit { path: path.to_string(), kind: f.kind.label(), len: f.len })
        .collect()
}

/// Walk parsed JSON, judging each scalar AND each key beside its scalar value.
///
/// FIELD-AWARE, never raw-line. A raw scan lets a digit run continue across JSON punctuation and
/// manufacture a card-shaped number out of two unrelated fields, which is what a first pass
/// appeared to find. The key/value pass is the other half: `{"api_key": "9f2b1c4d8e"}` has nothing
/// sensitive in either half alone, so a value-only walk reports it clean (Codex point 2).
fn walk(node: &serde_json::Value, path: &str, out: &mut Vec<Hit>) {
    match node {
        serde_json::Value::String(s) => out.extend(hits_of(path, s)),
        serde_json::Value::Number(n) => out.extend(hits_of(path, &n.to_string())),
        serde_json::Value::Array(a) => {
            for (i, v) in a.iter().enumerate() {
                walk(v, &format!("{path}[{i}]"), out);
            }
        }
        serde_json::Value::Object(map) => {
            for (k, v) in map {
                let child = if path.is_empty() { k.clone() } else { format!("{path}.{k}") };
                if let Some(scalar) = match v {
                    serde_json::Value::String(s) => Some(s.clone()),
                    serde_json::Value::Number(n) => Some(n.to_string()),
                    _ => None,
                } {
                    if let Some(f) = mind_types::sensitive_pair(k, &scalar) {
                        out.push(Hit { path: format!("{child} (key+value)"), kind: f.kind.label(), len: f.len });
                    }
                }
                walk(v, &child, out);
            }
        }
        _ => {}
    }
}

fn hits_of_json_line(line: &str) -> Vec<Hit> {
    let mut found = Vec::new();
    match serde_json::from_str::<serde_json::Value>(line) {
        Ok(v) => walk(&v, "", &mut found),
        // A line that will not parse is scanned as text — assuming a schema that does not apply
        // would report a clean bill of health for a file the audit never looked inside.
        Err(_) => found.extend(hits_of("<unparsed line>", line)),
    }
    found
}

struct Report {
    scanned: usize,
    per_kind: BTreeMap<&'static str, usize>,
    flagged: Vec<(String, Vec<Hit>)>,
}

impl Report {
    fn new() -> Self {
        Report { scanned: 0, per_kind: BTreeMap::new(), flagged: Vec::new() }
    }
    fn record(&mut self, id: String, hits: Vec<Hit>) {
        self.scanned += 1;
        if hits.is_empty() {
            return;
        }
        let mut kinds: Vec<&'static str> = hits.iter().map(|h| h.kind).collect();
        kinds.sort_unstable();
        kinds.dedup();
        for k in kinds {
            *self.per_kind.entry(k).or_insert(0) += 1;
        }
        self.flagged.push((id, hits));
    }
    fn print(&self, path: &str, unit: &str, explain: bool) {
        println!("{path}");
        println!("  {unit} scanned : {}", self.scanned);
        println!("  {unit} flagged : {}", self.flagged.len());
        for (k, n) in &self.per_kind {
            println!("    {k:22} {n}");
        }
        for (id, hits) in self.flagged.iter().take(20) {
            let mut kinds: Vec<&str> = hits.iter().map(|h| h.kind).collect();
            kinds.sort_unstable();
            kinds.dedup();
            println!("    id={id} kinds={}", kinds.join(","));
            if explain {
                // Key path and span LENGTH only. This is how the chain-hash false positives were
                // identified without anyone reading one.
                for h in hits {
                    println!("        at {} kind={} span_len={}", h.path, h.kind, h.len);
                }
            }
        }
        if self.flagged.len() > 20 {
            println!("    … {} more flagged records not listed", self.flagged.len() - 20);
        }
        println!();
    }
}

fn audit_jsonl(path: &str) -> std::io::Result<Report> {
    use std::io::BufRead;
    let f = std::fs::File::open(path)?;
    let mut rep = Report::new();
    for (i, line) in std::io::BufReader::new(f).lines().enumerate() {
        let line = match line {
            Ok(l) => l,
            // Invalid UTF-8 is itself worth seeing rather than skipping silently.
            Err(_) => {
                rep.record(format!("line:{}", i + 1), vec![Hit { path: "<invalid utf-8>".into(), kind: "unreadable-line", len: 0 }]);
                continue;
            }
        };
        if line.trim().is_empty() {
            continue;
        }
        rep.record(format!("line:{}", i + 1), hits_of_json_line(&line));
    }
    Ok(rep)
}

fn audit_db(path: &str) -> Result<Report, String> {
    let conn = rusqlite::Connection::open_with_flags(
        path,
        rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY | rusqlite::OpenFlags::SQLITE_OPEN_URI,
    )
    .map_err(|e| e.to_string())?;
    let mut stmt = conn.prepare("SELECT rid, text FROM memories").map_err(|e| e.to_string())?;
    let rows = stmt
        .query_map([], |r| Ok((r.get::<_, String>(0)?, r.get::<_, Option<String>>(1)?.unwrap_or_default())))
        .map_err(|e| e.to_string())?;
    let mut rep = Report::new();
    for row in rows {
        let (rid, text) = row.map_err(|e| e.to_string())?;
        rep.record(rid, hits_of("memories.text", &text));
    }
    Ok(rep)
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let fail_closed = args.iter().any(|a| a == "--fail-closed");
    // Key paths and span lengths for each finding — metadata about a value, never the value.
    let explain = args.iter().any(|a| a == "--explain");
    let paths: Vec<&String> = args.iter().filter(|a| !a.starts_with("--")).collect();

    println!("E.SEC1b READ-ONLY AUDIT — production's detector; counts, kinds and record ids only\n");

    // The instrument proves itself BEFORE it is trusted to report anything.
    match control_passes() {
        Ok(()) => println!("sensitivity control: PASS ({} canaries flagged, {} clean controls passed)\n", CANARIES.len(), CLEAN.len()),
        Err(why) => {
            println!("sensitivity control: FAIL — {why}");
            println!("\nINCONCLUSIVE. This run proves nothing about the files it was pointed at.");
            std::process::exit(2);
        }
    }

    let mut any = false;
    for path in paths {
        if !std::path::Path::new(path).exists() {
            println!("{path}: absent\n");
            continue;
        }
        let rep = if path.ends_with(".jsonl") {
            match audit_jsonl(path) {
                Ok(r) => {
                    r.print(path, "lines", explain);
                    r
                }
                Err(e) => {
                    println!("{path}\n  unreadable: {e}\n");
                    continue;
                }
            }
        } else {
            match audit_db(path) {
                Ok(r) => {
                    r.print(path, "memories", explain);
                    r
                }
                Err(e) => {
                    println!("{path}\n  unreadable: {e}\n");
                    continue;
                }
            }
        };
        any |= !rep.flagged.is_empty();
    }

    if any && fail_closed {
        println!("FINDINGS PRESENT — exiting non-zero (fail-closed).");
        std::process::exit(1);
    }
}
